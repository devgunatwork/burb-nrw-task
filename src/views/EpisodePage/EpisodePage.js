import React, { useEffect } from 'react';
import { useParams } from 'react-router-dom';
import Typography from '@material-ui/core/Typography';
import { useDispatch, useSelector } from 'react-redux';
import Error from '../../components/Error';
import Loader from '../../components/Loader';
import Details from '../../components/Details';
import SplitPageLayout from '../../components/SplitPageLayout';
import checkImageUrl from '../../utils/checkImageUrl';
import { FAILED, LOADING } from '../../constants/apiState.constants';
import { EPISODE_FETCH_REQUESTED } from '../../state/types/episodeSaga.type';

/**
 * @description component is container for episode detail page
 * */
const EpisodePage = () => {
  const { details, status: episodeRequestStatus } = useSelector((state) => state.episode);
  const { image, name, summary } = details;
  const url = checkImageUrl(image);
  const { episodeId } = useParams();
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch({ type: EPISODE_FETCH_REQUESTED, payload: { episodeId } });
  }, [episodeId, dispatch]);

  if (episodeRequestStatus === LOADING) return <Loader />;
  if (episodeRequestStatus === FAILED) return <Error message="Failed to fetch episode details" />;
  return (
    <SplitPageLayout imageUrl={url}>
      <Typography variant="h5">EPISODE DETAILS</Typography>
      <Details name={name} summary={summary} id={episodeId} />
    </SplitPageLayout>
  );
};

export default EpisodePage;
