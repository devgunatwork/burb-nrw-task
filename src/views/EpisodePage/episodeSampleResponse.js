export default {
  id: 1662010,
  url: 'http://www.tvmaze.com/episodes/1662010/the-powerpuff-girls-3x39-sideline-dad',
  name: 'Sideline Dad',
  season: 3,
  number: 39,
  type: 'regular',
  airdate: '2019-06-16',
  airtime: '17:30',
  airstamp: '2019-06-16T21:30:00+00:00',
  runtime: 15,
  image: {
    original: 'http://static.tvmaze.com/uploads/images/original_untouched/127/319448.jpg',
  },
  summary: '<p>When the Girls start playing soccer, the Professor goes overboard.</p>',
  _links: { self: { href: 'http://api.tvmaze.com/episodes/1662010' } },
};
