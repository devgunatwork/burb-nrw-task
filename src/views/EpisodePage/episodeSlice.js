import { createSlice } from '@reduxjs/toolkit';
import { FAILED, IDLE, LOADING, SUCCEEDED } from '../../constants/apiState.constants';

export const initialState = {
  details: {},
  status: IDLE,
  error: '',
};

export const episodeSlice = createSlice({
  name: 'episode',
  initialState,
  reducers: {
    requested: (state) => {
      state.status = LOADING;
    },
    succeeded: (state, action) => {
      state.status = SUCCEEDED;
      state.details = action.payload.episode;
    },
    failed: (state, action) => {
      state.status = FAILED;
      state.error = action.error;
    },
  },
});

export const { requested, succeeded, failed } = episodeSlice.actions;

export default episodeSlice.reducer;
