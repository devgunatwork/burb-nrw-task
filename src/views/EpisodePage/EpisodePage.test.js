import React from 'react';
import { useSelector } from 'react-redux';
import { initialState as episodeInitalState } from './episodeSlice';
import EpisodePage from './EpisodePage';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: () => ({
    episodeId: '1234',
  }),
}));
jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

describe('views', () => {
  describe('EpisodePage', () => {
    describe('UI', () => {
      it('should match the snapshot', () => {
        useSelector.mockImplementation((selector) => selector({ episode: episodeInitalState }));
        const wrapper = shallow(<EpisodePage />);

        expect(wrapper).toMatchSnapshot();
      });
    });
  });
});
