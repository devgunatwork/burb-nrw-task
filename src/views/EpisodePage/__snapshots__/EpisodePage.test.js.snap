// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`views EpisodePage UI should match the snapshot 1`] = `
<SplitPageLayout
  imageUrl="/assets/images/404.png"
>
  <WithStyles(ForwardRef(Typography))
    variant="h5"
  >
    EPISODE DETAILS
  </WithStyles(ForwardRef(Typography))>
  <Details
    id="1234"
    name=""
    summary=""
  />
</SplitPageLayout>
`;
